File: common/src/main/java/com/sumian/common/utils/ImageLoader.java
package com.sumian.common.utils;

import android.app.Activity;
import android.graphics.drawable.Drawable;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.widget.ImageView;

import com.bumptech.glide.RequestBuilder;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.DrawableImageViewTarget;

import de.hdodenhof.circleimageview.CircleImageView;

/**
 * Glide 图片加载辅助类
 * 适配圆形图片加载情况
 */

public class ImageLoader {
    private ImageLoader() {
    }

    public static void loadImage(Activity activity, ImageView view, String url, int placeholder) {
//        loadImage(Glide.with(activity), view, url, placeholder, placeholder);
    }

    public static void loadImage(Activity activity, ImageView view, String url, int placeholder, int error) {
//        loadImage(Glide.with(activity), view, url, placeholder, error);
    }

    public static void loadImage(Fragment fragment, ImageView view, String url, int placeholder) {
//        loadImage(Glide.with(fragment), view, url, placeholder, placeholder);
    }

    public static void loadImage(Fragment fragment, ImageView view, String url, int placeholder, int error) {
//        loadImage(Glide.with(fragment), view, url, placeholder, error);
    }

    public static void loadImage(RequestManager loader, ImageView view, String url) {
//        loadImage(loader, view, url, 0);
    }

    private static void loadImage(RequestManager loader, ImageView view, String url, int placeholder) {
//        loadImage(loader, view, url, placeholder, placeholder);
    }

    private static void loadImage(RequestManager loader, ImageView view, String url, int placeholder, int error) {
        boolean isCenterCrop = false;
        if (view instanceof CircleImageView) {
            isCenterCrop = true;
        }
        loadImage(loader, view, url, placeholder, error, isCenterCrop);
    }

    private static void loadImage(RequestManager loader, final ImageView view, String url, int placeholder, int error, boolean isCenterCrop) {
        if (TextUtils.isEmpty(url)) {
            view.setImageResource(placeholder);
        } else {
            RequestOptions options = RequestOptions.placeholderOf(placeholder).error(error).diskCacheStrategy(DiskCacheStrategy.AUTOMATIC);
            if (isCenterCrop) {
                options = options.centerCrop();
            }

            if (view instanceof CircleImageView) {
                RequestBuilder<Drawable> builder = loader.load(url).apply(options);
                builder.into(new DrawableImageViewTarget(view));
            } else {
                RequestBuilder builder = loader.load(url).apply(options);
                builder.into(view);
            }
        }
    }
}

File: common/src/main/java/com/sumian/common/image/ImageLoader.kt
package com.sumian.common.image

import android.view.View
import android.widget.ImageView
import androidx.annotation.DrawableRes
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions


/**
 * Created by sm
 *
 *
 * on 2018/8/1
 *
 *
 * desc:图片加载,因为有各种默认值,因此可以重载参数列表
 */
class ImageLoader {

    companion object {

        @JvmOverloads
        @JvmStatic
        fun loadImage(url: String, iv: ImageView, @DrawableRes placeHolderId: Int = -1, @DrawableRes errorId: Int = -1, isCenterCrop: Boolean = false) {

            var options = RequestOptions.placeholderOf(placeHolderId).error(errorId).diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)

            if (isCenterCrop) {
                options = options.centerCrop()
            }

            if (iv.visibility != View.VISIBLE) {
                iv.visibility = View.VISIBLE
            }

            Glide.with(iv.context).load(url).apply(options).into(iv)
        }

        @JvmOverloads
        @JvmStatic
        fun loadImage(@DrawableRes drawableId: Int = -1, iv: ImageView, @DrawableRes placeHolderId: Int = -1, @DrawableRes errorId: Int = -1, isCenterCrop: Boolean = false) {

            var options = RequestOptions.placeholderOf(placeHolderId).error(errorId).diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)

            if (isCenterCrop) {
                options = options.centerCrop()
            }

            if (iv.visibility != View.VISIBLE) {
                iv.visibility = View.VISIBLE
            }

            Glide.with(iv.context).load(drawableId).apply(options).into(iv)
        }

    }
}

/**
 * 拓展函数  可以更加方便的使用并加载图片   用法为 kotlin: ImageView().load()
 *                                             java:  ImageLoaderKt.load()
 *
 * @receiver ImageView
 * @param drawableId Int
 * @param placeHolderId Int
 * @param errorId Int
 * @param isCenterCrop Boolean
 */
@JvmOverloads
fun ImageView.loadImage(@DrawableRes drawableId: Int = -1, @DrawableRes placeHolderId: Int = -1, @DrawableRes errorId: Int = -1, isCenterCrop: Boolean = false) {

    var options = RequestOptions.placeholderOf(placeHolderId).error(errorId).diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)

    if (isCenterCrop) {
        options = options.centerCrop()
    }

    if (visibility != View.VISIBLE) {
        visibility = View.VISIBLE
    }
    Glide.with(context).load(drawableId).apply(options).into(this)
}

/**
 * 拓展函数  可以更加方便的使用并加载图片   用法为 kotlin: ImageView().load()
 *                                             java:  ImageLoaderKt.load()
 * @receiver ImageView
 * @param url String
 * @param placeHolderId Int
 * @param errorId Int
 * @param isCenterCrop Boolean
 */
@JvmOverloads
fun ImageView.loadImage(url: String, @DrawableRes placeHolderId: Int = -1, @DrawableRes errorId: Int = -1, isCenterCrop: Boolean = false) {

    var options = RequestOptions.placeholderOf(placeHolderId).error(errorId).diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)

    if (isCenterCrop) {
        options = options.centerCrop()
    }

    if (visibility != View.VISIBLE) {
        visibility = View.VISIBLE
    }
    Glide.with(context).load(url).apply(options).into(this)
}
